### ISSUE TITLE
Add proxy HTTPChannel object to allow swapping of channel after connection establishment.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8193]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8193|
|Type|enhancement|
|Created|2016-02-02 13:19:15Z|
|Branch|https://github.com/twisted/twisted/tree/httpchannel-proxy-8193-3|

This is one of a series of patches needed for HTTP/2 support: see [#7460](https://github.com/twisted/twisted/issues/7460) for more.

This patch adds a `GenericHTTPChannel` proxy object that is used to proxy calls through to the currently-existing `HTTPChannel` object. This exists to allow twisted.web to respond to upgrades made via NPN/ALPN by swapping out the `HTTPChannel` for a HTTP/2 equivalent.

This object is fairly simple in its approach, with one little caveat: twisted.web likes to do setup on its HTTPChannel object before it is used. This object needs to keep track of the changes made to the HTTPChannel to apply them to the HTTP/2 equivalent if it needs to be swapped in.

It's not immediately apparent how best to test this in isolation, but I'd welcome feedback on how to arrange some testing for this.

Attachments:

* [8193_1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/962/9626db05e03261ebdd243d19fccdbd858e6f895e/ffc2f9c873d94ca1922fcc8d6146251862db526c.patch) (3376 bytes) - added by Lukasa on 2016-02-02 13:20:15Z - First draft patch
* [8193_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/962/9626db05e03261ebdd243d19fccdbd858e6f895e/b2d93ed30106316eed290db15e873c4991e69d7b.patch) (5866 bytes) - added by Lukasa on 2016-02-15 10:39:05Z - Second draft patch, using proxyForInterface
* [8193_3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/962/9626db05e03261ebdd243d19fccdbd858e6f895e/65218ce56a5c2f4186193efec4f6cee242bb84ea.patch) (5877 bytes) - added by Lukasa on 2016-03-15 09:59:29Z - Third draft patch
* [8193_4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/962/9626db05e03261ebdd243d19fccdbd858e6f895e/4358de8ec307b8069e71616cb72e384ede8a09e9.patch) (9238 bytes) - added by Lukasa on 2016-03-15 14:02:40Z - Fourth draft patch

<details><summary>Searchable metadata</summary>

```
trac-id__8193 8193
type__enhancement enhancement
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__branches_httpchannel_proxy_8193_3 branches/httpchannel-proxy-8193-3
branch_author__hawkowl__adiroiban hawkowl, adiroiban
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1454419155233580 1454419155233580
changetime__1458055120312023 1458055120312023
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

I've uploaded the first patch, so I'm making this available for review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @Lukasa|
|-|-|

Thanks, HTTP2, small patches, etc. OK.

I would much rather define an explicit `IHTTPChannel` interface, and then use https://twistedmatrix.com/documents/15.5.0/api/twisted.python.components.html#proxyForInterface to construct the proxy, then add more code in Twisted for doing `__getattr__` hackery.  What do you think?

Also, I'd like to be careful about exposing any more public names that we have to support as a compatibility contract.  Is `GenericHTTPChannel` a thing that users are going to need to be able to construct?

I hope that this sort of high-level feedback is useful; given the nature of these patches I am reviewing them more from a high-level architectural perspective than the usual 'make sure that code we want to land is of acceptable quality' overview.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_3"></a>|@Lukasa commented|
|-|-|

> I would much rather define an explicit IHTTPChannel interface

This might be my inexperience with Twisted, but what does such an interface look like?

`HTTPChannel` currently defines 5 public class variables, one public instance variable, and 10 functions directly, but it subclasses `LineReceiver` which adds three more class variables and 6 more functions, and `TimeoutMixin` which defines one new class variable and four new functions. Now, that's a pretty large interface, but `H2Connection` (the nearest analogue to `HTTPChannel` in this instance) does *not* define all those (in particular it does not subclass `LineReceiver` for obvious reasons), which means that `H2Connection` would itself not conform to that interface.

Now, we *probably* don't need all those functions and class variables on the interface. In practice we're only proxying the protocol side of things, I think, so we maybe only need to proxy out IProtocol? But I'm kinda in the dark as to the best way to proceed with this, so some guidance would be helpful I think.

This means that I'm not really sure how to construct the interface you want here. I don't object to using it at all, but I'll need a bit of guidance as to what that actually looks like in practice.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_4"></a>|@Lukasa commented|
|-|-|

After chatting with glyph on IRC, we've come to some consensus around `proxyForInterface`. Specifically, we want to construct a `_GenericHTTPChannelProtocol` using `proxyFromInterface(IProtocol)` that has a mutable original value (glyph has requested that we don't call the attribute `original`, which seems reasonable). This solves the abstraction problem without me writing all those __getattr__s.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_5"></a>|@Lukasa commented|
|-|-|

Argh, so actually `proxyForInterface` may not be suitable. One extra thing my code did that `proxyForInterface` does not do is replay the changes made to the proxied object when it's swapped out. This works because things in twisted.web like to reach into the `HTTPChannel` and set some things (e.g. `requestFactory`).

I may be able to work around this by making `requestFactory` a property of the proxy object, but at a certain point we're patching around a lot of what twisted.web expects to do.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_6"></a>|@Lukasa removed owner|
|-|-|

Ok, I've provided a new patch.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

(In [46813]) Branching to httpchannel-proxy-8193.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

Related to my comment from [#8191](https://github.com/twisted/twisted/issues/8191)

Maybe we could use this new interface to also support a StringTransport backed HTTPChannel so that the Request object will only have to talk with an IHTTPChannel and don't have to also care about the transport.

With the current design, for me it make sense to start with a HTTPChannelWithStringTransport and later upgrade it to a File/Stream based HTTPChannel.

Not sure if this is useful or related to this ticket.

---------

1. Do we need to include the _negotiatedProtocol attribute and dataReceived together with this patch? The ticket description only talks about allowing to swap the HTTPhannel

2. Why do we need _requestFactory and _site? It is not obvious for me.

3. Is ``` H2_ENALED = False ```  related to this ticket?

I guess that this patch should include some example code or some more documentation.

By reading this patch, I am not sure how the actual swapping will be done.

---------

Patch applied for GitHub diff... and not sent to buildbot.

Leaving for review so that other can send their feedback.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_9"></a>|@Lukasa commented|
|-|-|

> Why do we need _requestFactory and _site? It is not obvious for me.

The `Site` class sets them. Lots of `twisted.web` seems to know quite a lot about other classes. ;)

> Is H2_ENALED = False related to this ticket?

It's a placeholder for the ticket that actually enables HTTP/2. Basically, that line will change.

> By reading this patch, I am not sure how the actual swapping will be done.

This patch *does* the swapping. If you look at `dataReceived`, you'll see that it potentially changes what `self._obj` is: specifically, changing it to a different object.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set owner to @Lukasa|
|-|-|

Hi Lukasa, thanks for the patch.

A few items:

- I super dislike ```_obj```, I don't find it's at all descriptive. ```_channel``` might make more sense?

- The docs for _obj says that it "conform[s] to the HTTPChannel protocol" but the interface only implements IProtocol. Hence, it should say that, but also say the type is an object implementing IProtocol's interface.

- References to Python objects should be wrapped in ```L{```} so that links can be made. In the datarecv function, you'd want to reference ```IProtocol.dataReceived```.

- A topfile is required. This could probably be a .misc for right now.

- Your reference to H2Connection doesn't exist, and hence makes the pyflakes checker sad (https://buildbot.twistedmatrix.com/builders/pyflakes/builds/3550/steps/pyflakes/logs/new%20pyflakes%20errors ). I'd prefer if that branch instead raised an assertion error, and it was put back in the patch that includes H2Connection.

Please fix these items and submit for rereview. I think it's close :)

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_11"></a>|@Lukasa removed owner|
|-|-|

Thanks for the review hawkowl. I've made the changes you suggested, and uploaded the third draft patch. Placing this back in the review queue.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

(In [46995]) Branching to httpchannel-proxy-8193-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

(In [46996]) apply patch from lukasa, refs [#8193](https://github.com/twisted/twisted/issues/8193)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl set owner to @Lukasa|
|-|-|

Hi Lukasa, thanks again for the updated patch.

As mentioned on IRC, there's some branch coverage issues in the new code -- although these would be covered when the rest of the H2 stuff lands, I would like some direct unit tests for that code before it lands -- and it won't hurt when the rest of H2 support lands, anyway.

Thanks again! Please submit for rereview once you've done this, and I think we're good to land.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_15"></a>|@Lukasa removed owner|
|-|-|

Ok, I've uploaded a new patch with tests that validate the behaviour of the class for various kinds of negotiation. I've also changed the negotiation code somewhat to confirm that branches that should not ever be reached are asserted as such.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl commented|
|-|-|

(In [46997]) Branching to httpchannel-proxy-8193-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl commented|
|-|-|

(In [46998]) apply patch from lukasa, refs [#8193](https://github.com/twisted/twisted/issues/8193)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl set owner to @hawkowl|
|-|-|

Yeah, I'm happy with this. Will land. Thanks for the work, Cory :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46999]) Merge httpchannel-proxy-8193-3: Add proxy HTTPChannel object to allow swapping of channel after connection establishment

Author: lukasa
Reviewers: glyph, adiroiban, hawkowl
Fixes: [#8193](https://github.com/twisted/twisted/issues/8193)
